[Cleanup] Linux build warnings for libfoundation, libscript, and toolchain

peter-b opened the pull request:
Includes a couple of bug fixes.

peter-b committed:
libfoundation: Clean up signed/unsigned comparisons.

Try to ensure that numeric comparisons in libfoundation are always
performed between numbers of the same signedness and, wherever
possible, the exact same representation.  The most common case is a
loop from 0 to a limit (stored in a variable), in which case the loop
variable should be the same type as the limit variable.

peter-b committed:
libscript: Clean up unused variable.

peter-b committed:
libfoundation: Add disambiguating parentheses.

peter-b committed:
libfoundation: Clean up C++ class member construction ordering

peter-b committed:
libfoundation: Clean up unreachable code detection.

Use `MCUnreachableReturn()` in a switch, instead of letting it fall
through to an `Assert(false)`.

peter-b committed:
libfoundation: Indirect MCValue access corrections.

In several libfoundation value methods, an indirect value was
resolved, but the result of the resolution was ignored.  This could
cause incorrect behaviour.

peter-b committed:
libfoundation: Clean up unused variables.

peter-b committed:
libfoundation: Correct relative ! and & operator precedence.

`!` actually has higher precedence than `&`, so the following doesn't
do what you expect:

    !int_value & kBitFlag

peter-b committed:
lc-compile: Clean up signed/unsigned comparisons.

In particular, all loops over operand counts should use `unsigned
int`.

peter-b committed:
MCObject::load() shouldn't be called for MCStack instances

Make it impossible to call the `MCStack::load()` by making the base
class implementation private.

`MCStack` already provides an `MCStack::load()` override which causes
an assertion failure in debug mode.

peter-b committed:
lc-compile: Clean up unused variable.

peter-b committed:
Make "overloaded-virtual" warnings into errors.

All of the instances of this warning found so far have been bugs.
Where the behaviour is intentional, we should add `using` declarations
as necessary.

montegoulding commented on toolchain/lc-compile/src/set.c:
```diff
@@ -22,7 +22,7 @@ typedef unsigned int SetRef;
 
 #define MAX_SET_DEPTH 256
 SetRef s_sets[MAX_SET_DEPTH];
-unsigned int s_set_count = 0;
```
@peter-b I'm not sure why this should be signed here. Happy to be educated if you have the time. The rest of this looks ok for review to me but I don't want to review something ok when I don't understand it. After reviewing the code in context I'm even more perplexed that the APIs use signed parameters so maybe I'd better leave this review to someone that knows this bit of the platform.

peter-b replied:
@montegoulding If you look at the set API used in this file, you'll see that it uses `int` parameters throughout (for example, see `EnsureSetStackDepth()`.  Fairly often, this results in an `int` being compared with `s_set_count`.

When comparing an `int` to an `unsigned int`, the `int` is converted to the equivalent unsigned representation, bitwise.  For example, consider:

``` C++
int a = -1;
unsigned int b = 1;
bool c = (a < b);
```

In this case, `c = false` because -1 maps to `UINT_MAX`, which is more than 1.

There were two options here:
1. I could refactor the whole file to add overflow checks everywhere in order to ensure correctness.
2. I could change `s_set_count` to `int`

montegoulding replied:
Thanks @peter-b. I guess I was confused as to why the APIs used signed parameters as it wouldn't appear to be valid to use them with negative parameters.

runrevmark commented:
@livecode-vulcan e500443

runrevmark commented:
@livecode-vulcan review ok e500443

livecode-vulcan commented:
💙 review by @runrevmark ok e5004436bc2d1790adc5afbc7d5d84e8ed15d21e

livecode-vulcan commented:
😞 test failure e5004436bc2d1790adc5afbc7d5d84e8ed15d21e
- try-community-linux-x86: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/2706)
- try-community-mac: [failure](https://vulcan.livecode.com/builders/try-community-mac/builds/2716)
- try-community-linux-x86_64: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/2701)
- try-community-emscripten: [pending](https://vulcan.livecode.com/builders/try-community-emscripten/builds/2717)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/2723)
- try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/2729)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/2741)

peter-b committed:
Correctly override MCStack::kdown() in subclasses.

Both `MCPlayerVolumePopup::kdown()` and
`MCPlayerRatePopup::kdown()`, as implemented for OS X, incorrectly
overrode `MCStack::kdown()`, meaning that they would never
actually be called if either popup was sent key events via a
pointer typed as an `MCStack` or `MCObjcet`.

peter-b committed:
Correctly subclass std::exception in externalv1 API

The C++ `std::exception::what()` virtual method is declared
`const throw()`; in order to override it in a subclass, the
overriding method must have the same CV-qualifiers.

runrevmark commented:
@livecode-vulcan review ok f0f14c2

livecode-vulcan commented:
💙 review by @runrevmark ok f0f14c2628aac3a78de1177bc65083aedcadbc99

livecode-vulcan commented:
😞 test failure f0f14c2628aac3a78de1177bc65083aedcadbc99
- try-community-linux-x86: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/2709)
- try-community-mac: [failure](https://vulcan.livecode.com/builders/try-community-mac/builds/2719)
- try-community-linux-x86_64: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/2704)
- try-community-emscripten: [pending](https://vulcan.livecode.com/builders/try-community-emscripten/builds/2720)
- try-community-ios: [pending](https://vulcan.livecode.com/builders/try-community-ios/builds/2726)
- try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/2732)
- try-community-android-armv6: [pending](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/2744)

peter-b commented:
I simply removed the trailing commit, which was orthogonal to the other changes in this PR.
@livecode-vulcan review by @runrevmark ok 45a4420

livecode-vulcan commented:
💙 review by @runrevmark ok 45a442074859ba5b00d045e0abd615a3206ad2a0

livecode-vulcan commented:
😞 test failure 45a442074859ba5b00d045e0abd615a3206ad2a0
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/2711)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/2721)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/2706)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/2722)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/2728)
- try-community-win-x86: [failure](https://vulcan.livecode.com/builders/try-community-win-x86/builds/2734)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/2746)

peter-b committed:
lc-compile: Ensure variables are declared at start of function body

...because our C source code needs to be C89-compatible for the
benefit of the Visual Studio compiler.

peter-b commented:
Hopefully `INT_MAX` is now defined...
@livecode-vulcan retry 45a4420

livecode-vulcan commented:
😞 test failure 45a442074859ba5b00d045e0abd615a3206ad2a0
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/2722)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/2732)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/2717)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/2733)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/2739)
- try-community-win-x86: [failure](https://vulcan.livecode.com/builders/try-community-win-x86/builds/2745)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/2757)

livecodefraser commented:
@livecode-vulcan review okay c4158ef

livecode-vulcan commented:
💙 review by @livecodefraser ok c4158effee022592ee56747b714819c0797e311c

livecode-vulcan commented:
😎 test success c4158effee022592ee56747b714819c0797e311c
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/2724)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/2734)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/2719)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/2735)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/2741)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/2747)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/2759)
